pkg/validate/dynamic/dynamic.go

```
@@ -305,6 +311,95 @@ func (dv *dynamic) validateVnetPermissions(ctx context.Context, vnet azure.Resou
+	if err != nil {
+		return err
+	}
+
+	// we need to explicitly set the resource type to virtualNetworks/{vnetID}/subnets, as the
+	// ParseResourceID function does not properly handle parsing child resources (e.g.
+	// VNET = parent, subnet = child) and gives the incorrect resource type, effectively
+	// giving the incorrect resource ID which causes the validateActions method to fail.
+	subnetr.ResourceType = fmt.Sprintf("%s/%s/%s", vnetr.ResourceType, vnetr.ResourceName, "subnets")
+
+	errCode := api.CloudErrorCodeInvalidResourceProviderPermissions
+	if dv.authorizerType == AuthorizerClusterServicePrincipal {
+		errCode = api.CloudErrorCodeInvalidServicePrincipalPermissions
+	}
+
+	operatorName, err := dv.validateActions(ctx, &subnetr, []string{
+		"Microsoft.Network/virtualNetworks/subnets/join/action",
+		"Microsoft.Network/virtualNetworks/subnets/read",
+		"Microsoft.Network/virtualNetworks/subnets/write",
+	})
+
+	var noPermissionsErr *api.CloudError
+	if err != nil {
+		if dv.authorizerType == AuthorizerWorkloadIdentity {
+			noPermissionsErr = api.NewCloudError(
+				http.StatusBadRequest,
+				api.CloudErrorCodeInvalidWorkloadIdentityPermissions,
+				"",
+				fmt.Sprintf(
+					errMsgWIHasNoRequiredPermissionsOnSubnet,
+					*operatorName,
+					subnetr.String(),
+				))
+		} else {
+			noPermissionsErr = api.NewCloudError(
+				http.StatusBadRequest,
+				errCode,
+				"",
+				fmt.Sprintf(
+					errMsgSPHasNoRequiredPermissionsOnSubnet,
+					dv.authorizerType,
+					*dv.appID,
+					subnetr.String(),
+				))
+		}
+	}
+
+	if err == wait.ErrWaitTimeout {
+		return noPermissionsErr
+	}
+	if detailedErr, ok := err.(autorest.DetailedError); ok {
+		dv.log.Error(detailedErr)
+
+		switch detailedErr.StatusCode {
+		case http.StatusNotFound:
+			return api.NewCloudError(
+				http.StatusBadRequest,
+				api.CloudErrorCodeInvalidLinkedSubnet,
+				"",
+				fmt.Sprintf(
+					errMsgSubnetNotFound,
+					subnetr.String(),
+				))
```

I think we will need to "tolerate" missing subnets here and either silently continue, or do an explicit check beforehand on each subnet to ensure it exists before attempting to validate it - it is possible for a cluster to end up with missing or invalid subnets in these coordinates, simply by creating a MachineSet on the cluster that points to a nonexistent subnet (the MachineSet will fail to provision any machines, but that subnet will get propagated in the list of worker subnets evaluated here).

We've had a recent example of this happening on a production cluster, and have had to develop some fixes for the various scenarios this can cause (e.g. #4137). See this Jira epic for more context: https://issues.redhat.com/browse/ARO-15595

---

...though looking at the existing dynamic validation code more closely, we're already subject to validation "failures" in this scenario anyways, so we can consider not handling this issue in this PR if it proves to be too complex, and tackling the problem within the validation as a whole later on.

---

I agree that this issue could be handled across all of dynamic validation in a different PR.

---

Yeah, agreed.  If we are already subject to these failures, I think it makes the most sense to track separately, especially since this will branch out beyond just subnet object (e.g. route tables, nat gateways).